### DISCUSSION TITLE
LSP References: Links, Articles, Etc.

### DISCUSSION THREAD

**Discussion by tajmone** [pml-lang/pml-companion#42] [Language Server]
The thought of implementing a PML Language Server is quite intriguing, for it would provide a one-solution-for-all in terms of editors/IDEs support for PML (well, at least for any modern editor that support LSP).

But implementing a good Language Server does pose a number of challenges, both in term of error tolerance and speed. Error tolerance is required because a WIP document is rarely a valid PML document, but will contain main incomplete constructs. Speed is an important factor because the Language Server must be able to update the highlighted document at every edit (i.e. every keystroke), so a good Lang Server needs to find a balance between memory consumption and speed, via caching techniques and avoiding having to re-parse the document each time, but rather amend its AST as chunks are updated.

In [VSCode's documentation we read](https://code.visualstudio.com/api/language-extensions/language-server-extension-guide#error-tolerant-parser-for-language-server):

> ### Error Tolerant Parser for Language Server
>
> Most of the time, the code in the editor is incomplete and syntactically incorrect, but developers would still expect autocomplete and other language features to work. Therefore, an error tolerant parser is necessary for a Language Server: The parser generates meaningful AST from partially complete code, and the Language Server provides language features based on the AST.
>
> When we were improving PHP support in VS Code, we realized the official PHP parser is not error tolerant and cannot be reused directly in the Language Server. Therefore, we worked on [Microsoft/tolerant-php-parser][Tolerant PHP Parser] and left detailed notes that might help Language Server authors who need to implement an error tolerant parser.

Below I'm providing some useful links to LSP official docs, as well as some articles of interest. Hopefully, these might be a starting point to further investigate how a PML Lang Server could be implemented.

-------------------------------------------------------------------------------

# Reference Links

## LSP (Language Server Protocol)

- [Wikipedia » LSP]
- [LSP Home at MicroSoft] — official LSP/LSIF website:
    + [LSP Specification]
    + [LSP Implementations]:
        * [Language Servers]
        * [Supporting Tools]
        * [SDKs]
- [Langserver.org] — community-driven source of knowledge for LSP implementations:
    + [Language Servers](https://langserver.org/#implementations-server)
    + [LSP Clients](https://langserver.org/#implementations-client)

## LSIF (Language Server Index Format)

- [LSIF Home at MicroSoft]:
    + [LSIF Specification]

## JSON-RPC

- [Wikipedia » JSON-RPC]
- www.jsonrpc.org — official JSON-RPC website:
    + [JSON-RPC 2.0 Specification]

## Useful References on Syntax Parsing

Articles and resources of interest for creating LSP Lang Servers, even if the former might not directly refer to the LSP.

- _[Persistence, Facades and Roslyn's Red-Green Trees]_

### Error Tolerant Parsing

- [Tolerant PHP Parser] — GitHub repository:
    + [How it Works](https://github.com/Microsoft/tolerant-php-parser/blob/main/docs/HowItWorks.md)
- _[Constructing human-grade parsers]_ — Joe Groff's article on writing parsers that can recover from error and produce partial output.

### Incremental Parsing

- _[Incremental Packrat Parsing: The Secret To Fast Language Servers]_ — by Eric McCarthy:
    + [PDF version](https://ohmlang.github.io/pubs/sle2017/incremental-packrat-parsing.pdf)
    + [support material](https://ohmlang.github.io/sle17/)
- _[Packrat Parsing: a Practical Linear-Time Algorithm with Backtracking]_ — Bryan Ford's Master's Thesis, MIT.

[LSP Home at MicroSoft]: https://microsoft.github.io/language-server-protocol/
[LSP Specification]: https://microsoft.github.io/language-server-protocol/specifications/specification-current/
[LSP Implementations]: https://microsoft.github.io/language-server-protocol/implementors/servers/
[Language Servers]: https://microsoft.github.io/language-server-protocol/implementors/servers/
[Supporting Tools]: https://microsoft.github.io/language-server-protocol/implementors/tools/
[SDKs]: https://microsoft.github.io/language-server-protocol/implementors/sdks/

[Langserver.org]: https://langserver.org/
[Wikipedia » LSP]: https://en.wikipedia.org/wiki/Language_Server_Protocol

[Tolerant PHP Parser]: https://github.com/microsoft/tolerant-php-parser

[Constructing human-grade parsers]: https://duriansoftware.com/joe/constructing-human-grade-parsers

[Persistence, Facades and Roslyn's Red-Green Trees]: https://docs.microsoft.com/en-us/archive/blogs/ericlippert/persistence-facades-and-roslyns-red-green-trees

[Incremental Packrat Parsing: The Secret To Fast Language Servers]: https://unallocated.com/blog/incremental-packrat-parsing-the-secret-to-fast-language-servers/
[Packrat Parsing: a Practical Linear-Time Algorithm with Backtracking]: https://bford.info/pub/lang/thesis/

[LSIF Home at MicroSoft]: https://microsoft.github.io/language-server-protocol/overviews/lsif/overview/
[LSIF Specification]: https://microsoft.github.io/language-server-protocol/specifications/specification-current/

[Wikipedia » JSON-RPC]: https://en.wikipedia.org/wiki/JSON-RPC

[JSON-RPC 2.0 Specification]: https://www.jsonrpc.org/specification

**Comment by pml-lang** [comment]
Thanks for the links. Very useful.

**Comment by tajmone** [comment]
# Python textX

I've come across this interesting tutorial on a [textX], a Python tools to create custom DSLs, which is also capable of auto-producing a Lang Server for the DSL, as well as TextMate-like editor syntaxes:

- [Quick Domain-Specific Languages in Python with textX]

The above tutorial is specific to creating a Lang Server for VSCode, so it might be worth looking into it and see if it could be a viable option to handle PML 2.0.0.

- https://github.com/textX/textX
- https://textx.github.io/textX/



[Quick Domain-Specific Languages in Python with textX]: https://tomassetti.me/quick-domain-specific-languages-in-python-with-textx/

[textX]: https://pypi.org/project/textX/

**Reply by pml-lang** [reply]
Interesting!

And: textX is inspired by [Xtext](https://www.eclipse.org/Xtext/).

**Comment by tajmone** [comment]
# Must Read Article: _Modern Parser Generator_

This is a must read article for LSP, among other things:

- _[Modern Parser Generator]_ — by [Aleksey Kladov] (June, 2018).

It's a very interesting resume on the various parsing techniques and tools, their limitations, and insights into the challenges of modern editors and IDE relying on real parsers to provide advanced syntax info and functionality.

The author also provides some ideas about novel approaches, linking to experimental parsers which he wrote, etc.

The article poses a strong emphasis on the Rust language.



[Modern Parser Generator]: https://matklad.github.io/2018/06/06/modern-parser-generator.html

[Aleksey Kladov]: https://github.com/matklad "View Aleksey Kladov's GitHub profile"

**Reply by celtic-coder** [reply]
Hi Tristano (@tajmone),

The article is written by a member of the Rust programming language dev-tools team. So, I guess that it's not surprising the emphasis on Rust! 😄 

The article is certainly an in-depth analysis of the subject -- but it is another galaxy in terms of my understanding of these things. I applaud those who can handle this depth of information but, unfortunately, I'm not likely to be making much of a contribution in any technical sense. I would be glad to assist in other ways, if this is possible.

Kind Regards,
Liam

**Reply by tajmone** [reply]
> The article is written by a member of the Rust programming language dev-tools team.

Ah! didn't know that. Thanks for pointing it out.

> but it is another galaxy in terms of my understanding of these things.

Me too, but I persist in my struggle to dig the topic and, year after year, I'm gradually starting to get the picture.

I've been reading many articles, tutorials and books on the topic, and at least know I manage to understand at an abstract level what these terms mean, and their application. So far, I've managed to write a syntax scanner and analyzer (a Lexer), but still struggling with fully understanding how parsing work — but I won't give up, and I believe that eventually I'll have my _Eureka!_ moment where every bit of the puzzle falls in place and I can finally see the whole picture clearly. 

If you're interested in the topic, this is a very good book, freely readable online (book code and book contents sources on GitHub):

https://www.craftinginterpreters.com/

**Reply by celtic-coder** [reply]
The book **Crafting Interpreters** is insane! Bob Nystrom's own article entitled [Crafting "Crafting Interpreters"](http://journal.stuffwithstuff.com/2020/04/05/crafting-crafting-interpreters/) describes the process involved -- the amount of work was absolutely incredible!

I didn't realize that it was free online. Thanks for letting me know!
